File: solutions/prolog/kindergarten-garden/kindergarten_garden.pl
students([alice, bob, charlie, david, eve, fred, ginny, harriet, ileana, joseph, kincaid, larry]).

plants('C', clover).
plants('G', grass).
plants('R', radishes).
plants('V', violets).

garden(Garden, Child, Plants) :-
    students(Students),
    once(nth1(Number, Students, Child)), 
    split_string(Garden, "\n", "", RowStrings),
    maplist(string_chars, RowStrings, Rows),
    transform_rows_to_beds(Rows, Beds),
    nth1(Number, Beds, Bed),
    maplist(plants, Bed, Plants).

transform_rows_to_beds([[], []], []).
transform_rows_to_beds([[R1_1, R1_2 | Rest1], [R2_1, R2_2 | Rest2]], Beds) :-
    transform_rows_to_beds([Rest1, Rest2], RestOfBeds),
    Beds = [[R1_1, R1_2, R2_1, R2_2] | RestOfBeds].
    

File: solutions/fsharp/kindergarten-garden/KindergartenGarden.fs
module KindergartenGarden

open FSharpx.Text

type Plant =
    | Clover
    | Grass
    | Radishes
    | Violets
    | Nothing

let studentList =
    [ "Alice"
      "Bob"
      "Charlie"
      "David"
      "Eve"
      "Fred"
      "Ginny"
      "Harriet"
      "Ileana"
      "Joseph"
      "Kincaid"
      "Larry" ]

let getPlantType (plant: char) : Plant =
    match plant with
    | 'C' -> Clover
    | 'G' -> Grass
    | 'R' -> Radishes
    | 'V' -> Violets
    | _ -> failwith "Not a valid seed type"

let divideGardenPlots (plots: string) : string list =
    plots
    |> Strings.split '\n'
    |> Array.toList
    |> List.map (Seq.chunkBySize 2)
    |> List.map (Seq.toList)
    |> List.map (List.map System.String)
    |> (fun g -> List.map2 (fun a b -> a + b) g.[0] g.[1])

let plants (diagram: string) (student: string) : Plant list =
    let gardenPlots = divideGardenPlots diagram

    match List.tryFindIndex (fun x -> x = student) studentList with
    | Some s ->
        gardenPlots.[s]
        |> Seq.toList
        |> List.map getPlantType
    | None -> failwith "Not a student in this class"
